P2047-alloc-aware-optional.html

```diff
@@ -38,128 +38,70 @@
+do any allocation, and that the name is not appropriate
+- fixed the order of arguments in the <i>Returns</i> clause of what used to be <tt>alloc_optional</tt>
+- adjusted for LWG 2833
+- added discussion on deduction guides
+- removed mentions of using a type alias since the type alias approach has proven to be problematic
+- modified the specification of pmr::optional to cover both AA and non AA types
+- added CTAD specification
 
-<p>
-  The current definition of <tt>std::optional</tt> does not follow the rules
-  for an AA type, even when holding an AA value. This limitation
-  makes <tt>optional</tt> unsuitable for storage in an AA container when
-  memory allocation customization is needed.
-</p>
 
 <p>
-  In this paper, we propose a new, allocator-aware
-  <tt>std::pmr::optional</tt> usable as a container element type and in any
-  other context where allocator propagation is expected. This new type would
-  not replace the current <tt>std::optional</tt>, but would resolve to
-  <tt>std::optional</tt> when the value type is not AA. By automatically
-  handling the distinction between AA and non-AA value types,
-  <tt>pmr::optional</tt> can easily be used in generic code where it is not
-  known in advance if it would be storing an AA type or not.
+  This is a proposal for an allocator-aware optional type. For users
+  that use custom allocators, such a type is currently missing;
+  pair and tuple support allocators, but an optional type is also
+  deemed necessary, because some such users use optional types
+  as container elements. Therefore, an allocator-aware optional
+  type is necessary for the usual scoped-allocator use cases to work.
+<br/>  
+  This is a complete proposal. It has not been seen by LEWGI or LEWG yet.
 </p>
 
+<h2> Motivation</h2>
 <p>
-  This is a complete proposal with formal wording.
-  It has not yet been discussed in LEWGI or LEWG.
+  Current optional type does not retain the element's allocator when it is
+  disengaged. This is a problem when using pmr containers because there is
+  no way of ensuring that a pmr::vector&lt;optional&gt; has all of its
+  elements using the same allocator - if an optional object gets disengaged, 
+  the pmr container it belongs to can not ensure the same allocator will 
+  be used if the optional object is re-engaged.
+  We propose an optional type which ensures that all of its elements use the same allocator.  
+  The allocator is not part of the type in the sense that it's 
+  not a template parameter.
 </p>
 
-<h2>Motivation and Proposal Summary</h2>
 <p>
-  C++ has been moving towards a consistent PMR allocator model whereby
-  <em>allocator-aware (AA)</em> types adhere to a few rules:
+  In order to ensure its value object is always constructed with the same allocator, 
+  an allocator-aware optional needs to store the allocator used
+  for object creation. Storing an allocator incurs space overhead,
```

Comment:
Should we keep the explanation why it needs to store the allocator, i.e. the optional may become disengaged, and then re-engaged, and in the re-engagement case it'll (re-)use the allocator it was originally given?

Reply:
I think so. To be fair, I have no idea how it got lost in the first place. I didn't deliberately remove it. Will reinstate some of the missing bits.